lib/puppet/type/file.rb

```diff
@@ -719,7 +716,9 @@ def remove_existing(should)
   end
 
   def retrieve
-    if source = parameter(:source)
+    # `checksum_value` implies explicit management of all metadata, so skip metadata
+    # retrieval. Otherwise, if source is set, retrieve metadata for source.
+    if (source = parameter(:source)) && property(:checksum_value).nil?
```

Comment:
This may cause problems for people providing checksum_value on their own. However we haven't really resolved how permissions and ownership are updated, so it's not clear to me what the alternative would look like.